File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3049_earliest-second-to-mark-indices-ii/solutions/cpp/solution.cpp
#include <vector>
#include <unordered_map>
#include <queue>

using namespace std;

class Solution {
public:
    int earliestSecondToMarkIndices(vector<int>& nums, vector<int>& changeIndices) {
        int n = nums.size();
        int m = changeIndices.size();
        
        vector<int> timesToZero(n, INT_MAX); // Tracks the earliest time each index can be zeroed
        unordered_map<int, priority_queue<int, vector<int>, greater<int>>> map; // Maps index to the seconds it can be set to zero
        
        // Prepare the map of times for each index when it can be set to zero
        for (int i = 0; i < m; i++) {
            int idx = changeIndices[i] - 1; // Convert 1-based to 0-based
            map[idx].push(i + 1); // Push the second (1-based)
        }
        
        // Determine the minimum second each index can be zeroed
        for (int i = 0; i < n; i++) {
            if (!map[i].empty()) {
                timesToZero[i] = map[i].top(); // Earliest second we can zero this index
            }
        }
        
        // Process marking of indices, considering decrement operations
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq; // Min-heap: (time, index)
        
        for (int i = 0; i < n; i++) {
            if (nums[i] == 0) { // Already zero, can be marked from second 1
                pq.emplace(1, i);
            } else {
                pq.emplace(timesToZero[i], i); // Can potentially be zeroed later
            }
        }
        
        vector<bool> marked(n, false);
        int maxTime = 0;
        int marksCount = 0;
        
        while (!pq.empty()) {
            auto [time, idx] = pq.top();
            pq.pop();
            
            if (marked[idx]) continue; // If already marked, skip
            
            if (time == INT_MAX) { // If time is still INT_MAX, we cannot zero this index
                return -1;
            }
            
            // Mark the index
            marked[idx] = true;
            maxTime = max(maxTime, time); // Update the latest time we performed a marking
            marksCount++;
            
            if (marksCount == n) { // If all indices are marked
                return maxTime;
            }
        }
        
        return -1;
    }
};

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3049_earliest-second-to-mark-indices-ii/solutions/java/solution.java
import java.util.*;

class Solution {
    public int earliestSecondToMarkIndices(int[] nums, int[] changeIndices) {
        int n = nums.length;
        int m = changeIndices.length;
        
        // Track minimum seconds required to make each index zero
        int[] earliestZero = new int[n + 1];
        Arrays.fill(earliestZero, Integer.MAX_VALUE);

        // Adjust indices in changeIndices for 1-indexed array
        for (int i = 0; i < m; i++) {
            changeIndices[i]--;
        }

        // Track the count of decrements still needed for each index
        int[] decrementsNeeded = nums.clone();
        
        // Frequency count of how many times we can set each index to zero directly
        int[] canSetZero = new int[n];
        
        // For each second, update the capabilities
        for (int i = 0; i < m; i++) {
            int index = changeIndices[i];
            canSetZero[index]++;
            
            // If at second i+1 we can set the value to 0, calculate when it's possible to mark this index
            if (decrementsNeeded[index] <= canSetZero[index]) {
                earliestZero[index] = Math.min(earliestZero[index], i + 1);
            }
        }
        
        // Check the possibility of marking all indices
        int latestSecond = 0;
        for (int i = 0; i < n; i++) {
            if (earliestZero[i] == Integer.MAX_VALUE) {
                return -1;
            }
            latestSecond = Math.max(latestSecond, earliestZero[i]);
        }
        
        return latestSecond;
    }
}
